### PR TITLE
[RCF] Dynamically linked TA

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

File: talibs/ta/Android.mk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+LOCAL_PATH := $(call my-dir)
+
+local_module := 8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta
```
Comment (vchong) [inline talibs/ta/Android.mk]: Is this uuid correct? Or should it be 93038f64-871d-4773-b1f5-878697745eee.ta?

File: talibs/Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+export V?=0
+
+# If _HOST or _TA specific compilers are not specified, then use CROSS_COMPILE
+HOST_CROSS_COMPILE ?= $(CROSS_COMPILE)
```
Comment (jenswi-linaro) [inline talibs/Makefile]: In `optee_test/Makefile` it's `CROSS_COMPILE_HOST` instead. I think we should stick with one order.

Comment (jforissier) [reply talibs/Makefile]: Agreed, but I have just copied what other examples do so this should be addressed separately IMO.

File: talibs/ta/ta_main.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <stdio.h>
+#include <talib1.h>
+#include <talib2.h>
+#include <talib3.h>
+#include <tee_internal_api.h>
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	printf("TA_CreateEntryPoint()\n");
```
Comment (etienne-lms) [inline talibs/ta/ta_main.c]: `DMSG()`?

File: talibs/ta/user_ta_header_defines.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/* Provisioned stack size */
+#define TA_STACK_SIZE			(2 * 1024)
+
+/* Provisioned heap size for TEE_Malloc() and friends */
+#define TA_DATA_SIZE			(32 * 1024)
+
+/* Extra properties (give a version id and a string name) */
+#define TA_CURRENT_TA_EXT_PROPERTIES \
+    { "gp.ta.description", USER_TA_PROP_TYPE_STRING, \
+        "Example of OP-TEE Hello World Trusted Application" }, \
```
Comment (etienne-lms) [inline talibs/ta/user_ta_header_defines.h]: change description

File: talibs/talib1/talib1.c [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <stdio.h>
+#include "talib1.h"
+
+void talib1_func(void)
+{
+	printf("talib1_func()\n");
```
Comment (etienne-lms) [inline talibs/talib1/talib1.c]: `DMSG()`

File: talibs/host/main.c [hunk 6]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <err.h>
+#include <tee_client_api.h>
+#include <libs_ta.h>
+
+int main(int argc, char *argv[])
```
Comment (etienne-lms) [inline talibs/host/main.c]: could be move to xtest regression 1xxx?

Comment (jforissier) [reply talibs/host/main.c]: Well, for regression testing I'd rather refactor the storage TAs. Currently we have two TAs, `storage` and `storage2`, which differ only by their UUID. We could easily move all the common code into a shared library I suppose.

File: talibs/ta/user_ta_header_defines.h (hunk 4 above)
Comment (jforissier) [reply talibs/ta/user_ta_header_defines.h]: Will do

File: talibs/ta/ta_main.c (hunk 3 above)
Comment (jforissier) [reply talibs/ta/ta_main.c]: For sure `xMSG()` are convenient and are already used in other examples, but IMO `printf()` is simpler and does not depend on setting the proper CFG_TEE_TA_LOG_LEVEL flag.

Comment (etienne-lms) [reply talibs/ta/ta_main.c]: printf() is not referenced in the op-tee dev doc. So maybe removed one day.
So I think it's a bad idea in the **optee_examples** repo, even if quite convenient when hacking a bit.

File: talibs/host/main.c [hunk 7]
```diff
@@ -25,19 +25,13 @@ int main(int argc, char *argv[])
...
-			res, err_origin);
-
-	/*
-	 * We're done with the TA, close the session and
-	 * destroy the context.
-	 */
-
-	TEEC_CloseSession(&sess);
+	if (res != TEEC_ERROR_TARGET_DEAD)
+		errx(1, "TEEC_Opensession failed unexpectedly with code 0x%x "
```
Comment (etienne-lms) [inline talibs/host/main.c]: s/failed unexpectedly/returned unexpectedly/

File: talibs/ta/Makefile [hunk 8]
```diff
@@ -1,6 +1,6 @@
 LDADD = -L$(link-out-dir)/../talib1 -ltalib1 \
-	-L$(link-out-dir)/../talib2 -ltalib2 \
-	-L$(link-out-dir)/../talib3 -ltalib3
+	-L$(link-out-dir)/../talib2 -ltalib2 #\
+#	-L$(link-out-dir)/../talib3 -ltalib3
```
Comment (etienne-lms) [inline talibs/ta/Makefile]: ?

File: talibs/ta/ta_main.c (hunk 3 above)
Comment (etienne-lms) [reply talibs/ta/ta_main.c]: note: ok for `printf`

File: talibs/host/main.c [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	 * Open a session to the "library example" TA, the TA will invoke
+	 * functions implemented in libraries (libtalib1.a, libtalib2.so,
+	 * libtalib3.so). Watch the secure console for debug messages from the
+	 * libraries.
+	 * We expect a failure because the TA should call TEE_Panic().
+	 */
+	res = TEEC_OpenSession(&ctx, &sess, &uuid,
+			       TEEC_LOGIN_PUBLIC, NULL, NULL, &err_origin);
+	if (res != TEEC_ERROR_TARGET_DEAD)
+		errx(1, "TEEC_Opensession failed unexpectedly with code 0x%x "
```
Comment (etienne-lms) [inline talibs/host/main.c]: minor: s/failed unexpectedly/unexpectedly returned/

File: talibs/ta/Makefile (hunk 8 above)
Comment (jforissier) [reply talibs/ta/Makefile]: Sorry, I did a `git commit -a` by mistake. Only the change in `talibs/host/main.c` was intended. I'll fix that in a new commit.

**Pull request closed without merging**